feat: perform html filtering on headers received

seia-soto opened the pull request:
- Put `has-text` and `replace` filters accordingly by checking the conditions (including mime check)
- Check accumulated buffer size in streaming filter and abort if more than 10 megabytes is received.
- Move performHTMLFiltering to onHeadersReceieved to determine by the content-type header from server

seia-soto committed:
feat: perform html filtering on headers received

seia-soto commented:
refs https://github.com/ghostery/adblocker/pull/3897

seia-soto committed:
chore: drop duplicate check in getHTMLFilters

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,6 +131,43 @@ export function updateResponseHeadersWithCSP(
   return { responseHeaders };
 }
 
+const readableMimeTypes = new Set([
```
please rename to be full capitalised

seia-soto replied:
Fixed in [`658f616` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/658f6168de9d534d72b3b1db2e56a43fef716373)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -177,6 +218,11 @@ export function filterRequestHTML(
     }
 
+    accumulatedBufferSize += event.data.byteLength;
+    if (accumulatedBufferSize > MAXIMUM_RESPONSE_BUFFER_SIZE) {
```
please add unit test for this

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -405,9 +455,18 @@ export class WebExtensionBlocker extends FiltersEngine {
+      }
+      const replaceFilterIndex = htmlFilters.findIndex(([type]) => type === 'replace');
+      // If `replace` filters are not supported
+      if (isRequestHTMLFilterable(request, details) === false && replaceFilterIndex !== -1) {
```
how about moving this into `filterRequestHTML`?

seia-soto replied:
I think it's making props drilling too deep.

seia-soto committed:
chore: rename set of mime types

refs https://github.com/seia-soto/adblocker/pull/14/files#r1718049934

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -539,19 +598,14 @@ export class WebExtensionBlocker extends FiltersEngine {
     }
 
-    if (request.isMainFrame()) {
-      this.performHTMLFiltering(browser, request);
```
this cannot be simply removed - if any html filters are meant to be applied, they should prevent other filters from being run

chrmod replied:
I take this back. We should prioritise blocking

seia-soto committed:
chore: reuse `_originalRequestDetails` from `Request`

Co-authored-by: chrmod <chrmod@ghostery.com>

seia-soto committed:
chore: drop @ts-expect-error in `fromWebRequestDetails`

Co-authored-by: chrmod <chrmod@ghostery.com>

seia-soto committed:
chore: move logics into `filterRequestHTML`

Co-authored-by: chrmod <chrmod@ghostery.com>

seia-soto committed:
chore: revert unnecessary changes in 6b1bd2aab8f540f42cd4fadd192cf872350a300b

seia-soto committed:
chore: rename `isRequestHTMLFilterable`

Co-authored-by: chrmod <chrmod@ghostery.com>

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,64 @@ export function updateResponseHeadersWithCSP(
+]);
+const MAXIMUM_RESPONSE_BUFFER_SIZE = 10 * 1024 * 1024;
+
+export function shouldApplyReplaceSelectors(
```
please add unit tests for this

seia-soto committed:
feat: check content-length header

seia-soto committed:
test(adblocker-webextension): html-filtering

seia-soto committed:
chore: reject sending data on `disconnected`

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,213 @@ describe('#fromWebRequestDetails', () => {
+    ];
+
+    sets.forEach((set) => {
+      const details = createWebRequestDetails(set);
+      const request = fromWebRequestDetails(details);
```
please move this into `it` - otherwise those are initiated when tests are being prepared, not run, which makes tests selected by `only` to run this code.

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,74 @@ export function updateResponseHeadersWithCSP(
+  const contentTypeHeader = details.responseHeaders?.find(
+    (header) => header.name.toLowerCase() === 'content-type',
+  )?.value;
+  // Content-Length includes the length of octets which is consisted of 8 bytes with endianness
```
not sure if this comment is helping with anything

seia-soto replied:
Updated in [`bead6e9` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/bead6e965e01aab9d95de53b1f0c5f2efbdc3385)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,74 @@ export function updateResponseHeadersWithCSP(
+    (header) => header.name.toLowerCase() === 'content-type',
+  )?.value;
+  // Content-Length includes the length of octets which is consisted of 8 bytes with endianness
+  const contentLengthHeader = details.responseHeaders?.find(
+    (header) => header.name.toLowerCase() === 'content-length',
+  )?.value;
+  if (
+    contentLengthHeader !== undefined &&
+    parseInt(contentLengthHeader, 10) >= MAXIMUM_RESPONSE_BUFFER_SIZE
+  ) {
+    return false;
+  }
```
```suggestion
  if (
    parseInt(
      details.responseHeaders?.find(
        (header) => header.name.toLowerCase() === "content-length"
      )?.value ?? "0",
      10
    ) >= MAXIMUM_RESPONSE_BUFFER_SIZE
  ) {
    return false;
  }
```

bit more condensed

seia-soto replied:
Updated in [`bead6e9` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/bead6e965e01aab9d95de53b1f0c5f2efbdc3385)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,74 @@ export function updateResponseHeadersWithCSP(
+  request: Request,
+  details: OnHeadersReceivedDetailsType,
+): boolean {
+  const contentTypeHeader = details.responseHeaders?.find(
```
lets move this lookup below the content-length check

seia-soto replied:
Updated in [`bead6e9` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/bead6e965e01aab9d95de53b1f0c5f2efbdc3385)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,74 @@ export function updateResponseHeadersWithCSP(
   return { responseHeaders };
 }
 
+const READABLE_MIME_TYPES = new Set([
```
how about calling it `HTML_FILTERABLE_MIME_TYPES`?

seia-soto replied:
Updated in [`bead6e9` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/bead6e965e01aab9d95de53b1f0c5f2efbdc3385)

chrmod commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +132,74 @@ export function updateResponseHeadersWithCSP(
+
+export function shouldApplyReplaceSelectors(
+  request: Request,
+  details: OnHeadersReceivedDetailsType,
```
no need for a second argument, as the details are already a part of the Request type

seia-soto replied:
Updated in [`bead6e9` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/bead6e965e01aab9d95de53b1f0c5f2efbdc3385)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+      return details;
+    }
+
+    const sets: Array<Parameters<typeof createWebRequestDetails>[0] & { result: boolean }> = [
```
I'm not sure if this meta programming brings much. Can we replace it with a separate `it` blocks, each having a proper description?

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+      responseType: WebRequest.ResourceType;
+      headers: Record<string, string>;
+    }) {
+      const url = 'https://foo.com/';
+      const requestId = 'req-01';
+      const tabId = 1;
```
all those can be inlined with the details object literal

chrmod replied:
how about 
```js
const createWebRequestDetails = ({ type, headers }) => ({
  requestId: "req-01",
  tabId: 1,
  url: "https://foo.com/",
  type,
  responseHeaders: Object.entries(headers).map(([name, value]) => ({
    name,
    value,
  })),
});
```

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+
+    it('stops processing more than MAXIMUM_RESPONSE_BUFFER_SIZE', () => {
+      const streamFilter = new StreamFilter();
+      const details: OnHeadersReceivedDetailsType = {
+        responseHeaders: [
+          {
+            name: 'content-type',
+            value: 'text/text',
+          },
+        ],
+        url: 'https://foo.com/script.js',
+        type: 'main_frame',
+        originUrl: 'https://foo.com/',
+        tabId: 0,
+        requestId: 'req-00',
+      };
+      const request = fromWebRequestDetails(details);
```
can be merged together into 
```js
const request = fromWebRequestDetails({
  responseHeaders: [
    {
      name: 'content-type',
      value: 'text/text',
    },
  ],
  url: 'https://foo.com/script.js',
  type: 'main_frame',
  originUrl: 'https://foo.com/',
  tabId: 0,
  requestId: 'req-00',
});
```

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+        return streamFilter;
+      }
+
+      filterRequestHTML(filterResponseData, request, [['replace', [new RegExp('a', 'g'), 'b']]]);
```
please add comments on what this test is doing. here you can say 
```
// filter aims to replace every instance of letter `a` with letter `b` in the entire response body
```

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+        sum += octet;
+      }
+
+      streamFilter._reset();
```
I'm guessing that `reset` is not really needed as you create a new stream for each test

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+
+      streamFilter._push(Uint8Array.from(['b'.charCodeAt(0)]), true);
+
+      let sum = 0;
+      for (const octet of streamFilter._pulled) {
+        sum += octet;
+      }
```
what does this do?

seia-soto replied:
This adds all collected char codes of string, if there's 3 'a' characters, it'll result to the value of: 97 * 3. By doing this, we assume characters in the result are in wanted state.

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+
+      streamFilter._reset();
+
+      expect(sum).to.be.eql('b'.charCodeAt(0) * (MAXIMUM_RESPONSE_BUFFER_SIZE + 1));
```
I don't understand what this expectation is meant to mean. please add comments

seia-soto replied:
Updated in [`9991a4a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/9991a4a7979f149105991bd4b199be1250a86d41)

chrmod commented on packages/adblocker-webextension/test/index.test.ts:
```diff
@@ -125,3 +131,209 @@ describe('#fromWebRequestDetails', () => {
+      expect(sum).to.be.eql('b'.charCodeAt(0) * (MAXIMUM_RESPONSE_BUFFER_SIZE + 1));
+    });
+  });
+});
```
if this test is meant to prove that if content-length exceeds the MAX then no html filtering is executed, then we should add another then proves the html filters works if the content-length is just below the MAX

seia-soto replied:
We're already doing this, I updated the check to the following to make this more verbose:

```js
// We'll have NOISE_BUFFER_SIZE length of chunks which is replaced from letter 'a' to letter 'b'
// Then one letter 'c' which should not be replaced to letter 'a' or anything
expect(sum).to.be.eql('b'.charCodeAt(0) * MAXIMUM_RESPONSE_BUFFER_SIZE + 'c'.charCodeAt(0));
```

seia-soto replied:
Just an idea, we can check this more strictly by dividing the sum into MAXIMUM_RESPONSE_BUFFER_SIZE to check the length of output noises. I don't think this is worth here but if we're going to add more cases, additional check is preferred.

seia-soto committed:
test(adblocker-webextension): clean up

Co-authored-by: chrmod <chrmod@ghostery.com>

- Make #shouldApplyReplaceSelectors tests more verbose and simple
- Comments for #MAXIMUM_RESPONSE_BUFFER_SIZE tests

seia-soto committed:
test(adblocker-webextension): clean up

Co-authored-by: chrmod <chrmod@ghostery.com>

- Rename READABLE_MIME_TYPES to HTML_FILTERABLE_MIME_TYPES
- Extract details from request in shouldApplyReplaceSelectors
- Additional typing for Request to avoid typing mistakes in the future

seia-soto committed:
chore: add missing types for mocked StreamFilter

seia-soto committed:
feat: backward compatible `shouldApplyReplaceSelectors`

seia-soto committed:
chore: fix unnecessary check

chrmod committed:
Simplify request type detection

seia-soto commented:
@chrmod We're losing backward-compatibility here: [`f514f32` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/f514f32a314e1f3d9f208383cb79ef29388607ac). We need to set the type to `OnBeforeRequestDetails | OnHeadersReceivedDetails` instead.

seia-soto commented:
I get it. It's just simplifying the typings. Checked.

seia-soto committed:
test: improve readability

- Simpler check on output chunks of StreamFilterMock
- Improve readability on tests

Co-authored-by: philipp <philipp@ghostery.com>
Co-authored-by: chrmod <chrmod@ghostery.com>

seia-soto committed:
fix: import path

seia-soto committed:
chore: inline StreamFilterMock.isResumable

seia-soto committed:
chore: check content-disposition header

seia-soto committed:
test: content-disposition header check

seia-soto committed:
feat: check status code

chrmod committed:
Cleanup

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
   return { responseHeaders };
 }
 
+const HTML_FILTERABLE_MIME_TYPES = new Set([
```
`text/html`?

remusao replied:
I see `text/xxx` will be included by default below. Let's document this behavior here as well please.

seia-soto replied:
Leaving reference [`c71579a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/c71579a417157d5166b4efe58a8369aacd7baccb#diff-9545dc27a6748320d2b161f59633758bbfb8482494289e9181fcb5a85021b792R142)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+    ?.value;
+}
+
+export function shouldApplyReplaceSelectors(
```
Let's add more comments/documentation on this function please.

chrmod replied:
is this sufficient?

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+  }
+
+  if (
+    parseInt(getHeaderFromDetails(details, 'content-length') ?? '0', 10) >=
```
Let's use `Number` instead of `parseInt`.

seia-soto replied:
Leaving ref [`c71579a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/c71579a417157d5166b4efe58a8369aacd7baccb)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+    return false;
+  }
+
+  // If status code is non 2xx or 4xx, 5xx
+  const statusCodeScope = (details.statusCode / 100) >>> 0;
+  if (statusCodeScope !== 2 && statusCodeScope < 4) {
+    return false;
+  }
```
Don't we want to skip for any non-2xx code?
```suggestion
  // If status code is non 2xx or 4xx, 5xx
  if (details.statusCode < 200 || details.statusCode > 299) {
    return false;
  }
```

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+    return false;
+  }
+
+  if (getHeaderFromDetails(details, 'content-disposition')?.startsWith('inline') === false) {
```
Can you document what this does and when it can happen?

seia-soto replied:
Leaving ref [`c71579a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/c71579a417157d5166b4efe58a8369aacd7baccb#diff-9545dc27a6748320d2b161f59633758bbfb8482494289e9181fcb5a85021b792R183)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+
+  const contentTypeHeader = getHeaderFromDetails(details, 'content-type');
+  if (
+    contentTypeHeader !== undefined &&
```
Do we want to skip if there is no `content-type` available?

seia-soto replied:
Leaving ref [`c71579a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/c71579a417157d5166b4efe58a8369aacd7baccb#diff-9545dc27a6748320d2b161f59633758bbfb8482494289e9181fcb5a85021b792L194-R200)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+  const contentTypeHeader = getHeaderFromDetails(details, 'content-type');
+  if (
+    contentTypeHeader !== undefined &&
+    (contentTypeHeader?.startsWith('text') || HTML_FILTERABLE_MIME_TYPES.has(contentTypeHeader))
```
Should we lower-case `contentTypeHeader`?

seia-soto replied:
Leaving ref [`c71579a` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/c71579a417157d5166b4efe58a8369aacd7baccb#diff-9545dc27a6748320d2b161f59633758bbfb8482494289e9181fcb5a85021b792R197)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +139,98 @@ export function updateResponseHeadersWithCSP(
+    return true;
+  }
+
+  if (
+    !request.type.startsWith('image') &&
+    request.type !== 'font' &&
+    request.type !== 'websocket' &&
+    request.type !== 'media'
+  ) {
+    return true;
+  }
+
+  return false;
```
Would it make sense to instead include a list of types for which we do want to perform filtering? I think it'd be clearer.

seia-soto replied:
Leaving ref [`28d3fcd` (#14)](https://github.com/seia-soto/adblocker/pull/14/commits/28d3fcd26deb51d87b113a7d1f5645fc26023364)

remusao commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -564,13 +656,12 @@ export class WebExtensionBlocker extends FiltersEngine {
-      details,
-      this.getCSPDirectives(fromWebRequestDetails(details)),
-    );
+    const request = fromWebRequestDetails(details);
```
Do we need to specify the generic type `T` here?

chrmod replied:
from what I see it is being inferred from details

chrmod committed:
Update packages/adblocker-webextension/src/index.ts

Co-authored-by: Rémi <remusao@users.noreply.github.com>

chrmod committed:
Cleanup

chrmod committed:
Cleanup

seia-soto committed:
chore: drop `toLowerCase` in header name comparison

Latest firefox automatically lowers all header names.

```
import Fastify from 'fastify'
const fastify = Fastify({
  logger: true
})

// Declare a route
fastify.get('/', async function handler(request, reply) {
  reply.header('UPPERCASED', '1')
  return { hello: 'world' }
})

// Run the server!
try {
  await fastify.listen({ port: 3000 })
} catch (err) {
  fastify.log.error(err)
  process.exit(1)
}
```

seia-soto commented on packages/adblocker-webextension/src/index.ts:
```diff
@@ -128,19 +140,110 @@ export function updateResponseHeadersWithCSP(
+    return false;
+  }
+
+  if (details.statusCode < 200 || details.statusCode > 299) {
```
Shouldn't we handle 5xx and 4xx requests? 3xx are redirects but 5xx and 4xx still can include ad scripts.

chrmod replied:
I don't see why 4xx or 5xx would include ads - those are error codes. Do we have any proves that ads are shipped that way?
